src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -0,0 +1,415 @@
+            {
+                if (!_dagNodeLabels.TryGetValue(dag, out LabelSymbol label))
+                {
+                    _dagNodeLabels.Add(dag, label = dag is BoundDecision d ? d.Label : _factory.GenerateLabel("dagNode"));
+                }
+
+                return label;
+            }
+
+            /// <summary>
+            /// Lower the given nodes into _loweredDecisionDag.
+            /// </summary>
+            private void LowerDecisionDag(ImmutableArray<BoundDecisionDag> sortedNodes)
+            {
+                switch (sortedNodes[0])
+                {
+                    case BoundWhenClause wc:
+                        // If the first node is in a when clause's section rather than the code for the
+                        // lowered decision dag, jump there to start.
+                        _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(wc)));
+                        break;
+                    case BoundDecision d:
+                        // If the first node is a decision rather than the code for the
+                        // lowered decision dag, jump there to start.
+                        _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(d)));
+                        break;
+                }
+
+                // Call LowerDecisionDagNode with each node and its following node in the generation order.
+                // However, some nodes can be emitted more efficiently as a switch dispatch (possibly out
+                // of order).
+                var loweredNodes = PooledHashSet<BoundDecisionDag>.GetInstance();
+                BoundDecisionDag previous = null;
+                foreach (BoundDecisionDag node in sortedNodes)
+                {
+                    if (node is BoundWhenClause w)
+                    {
+                        // Code for the when clause goes in a separate code section for the switch section.
+                        LowerWhenClause(w);
+                        continue;
+                    }
+
+                    // BoundDecision nodes do not get any code generated for them.
+                    if (node is BoundDecision || loweredNodes.Contains(node))
+                    {
+                        continue;
+                    }
+
+                    if (previous != null && !loweredNodes.Contains(previous))
+                    {
+                        // Lower the node "previous".
+                        if (!GenerateSwitchDispatch(previous, loweredNodes))
+                        {
+                            LowerDecisionDagNode(previous, node);
+                        }
+
+                        loweredNodes.Add(previous);
+                    }
+
+                    previous = node;
+                }
+
+                // Lower the final node
+                if (previous != null && !loweredNodes.Contains(previous))
+                {
+                    LowerDecisionDagNode(previous, null);
```

Comment:
>LowerDecisionDagNode [](start = 20, length = 20)

Why don't we call `GenerateSwitchDispatch` for this case? Please add a comment, either to this case or the case inside the `foreach`, to describe the difference.

Reply:
Will add a comment, but the short answer is that we've reached the end of the set of nodes in the decision dag, having lowered them all, so there are no ungenerated following nodes that could possibly be part of a switch dispatch.

---
In reply to: [176305066](https://github.com/dotnet/roslyn/pull/25544#discussion_r176305066) [](ancestors = 176305066)

Reply:
Recorded in [https://github.com/dotnet/roslyn/issues/25663](https://github.com/dotnet/roslyn/issues/25663)

---
In reply to: [176312632](https://github.com/dotnet/roslyn/pull/25544#discussion_r176312632) [](ancestors = 176312632,176305066)